tox_core/src/net_crypto/mod.rs

```diff
@@ -1021,21 +1047,27 @@ impl NetCrypto {
 
     /// Send `CookieRequest` or `CryptoHandshake` packet if needed depending on
     /// connection status and update sent counter
-    fn send_status_packet(&self, connection: &mut CryptoConnection) -> impl Future<Output = Result<(), SendPacketError>> + Send {
-        match connection.packet_to_send() {
+    fn send_status_packet(&self, connection: Arc<RwLock<CryptoConnection>>, status: Option<ConnectionStatus>)
```

Comment:
Why did you change it everywhere to Arc<RwLock<..>>?

Reply:
I can explain it like:

1. `NetCrypto.handle_cookie_response()` is need to take ownership of Connection in `async {}` block, 
    can't do this by `&mut` because it pass connection to `send_status_packet()` across `await` point.
    That's, `send_status_packet()` should to mutate connection, but can't take &mut.
    My first try was an Arc<Mutex<, but.. 
2. `NetCrypto.main_loop()` is a "source" of `Connection`'s for another NetCrypto methods like `send_status_packet()`
3. `NetCrypto.connections` has type of `Arc<RwLock<HashMap<PublicKey, Arc<RwLock<CryptoConnection>>>>>`, 
    that means that `Arc<RwLock<CryptoConnection>>` is natural type for Connection in NetCrypto's architecture

Reply:
Another reason to migrate to `async fn` in this MR - this will allow to leave it as `&mut`.

Reply:
You can check https://github.com/tox-rs/tox/commit/0dd558fd1a6766a153311f80453a4f73cb07158e where I partly migrated to `async fn` - I haven't figured out how yet to make echo example work, but it should be possible :)